File: 1874/1874.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   1874.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/08/04 14:29:50 by jihoolee          #+#    #+#             */
/*   Updated: 2024/08/04 15:09:14 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <vector>
#include <stack>

std::vector<char> format_by_stack(const std::vector<int>& nums) {
  std::vector<char> operations;
  std::stack<int>   s;
  int               new_num = 1;

  for (size_t idx = 0; idx < nums.size(); ++idx) {
    while (s.empty() || s.top() != nums[idx]) {
      if (new_num > nums.size())
        return std::vector<char>();
      s.push(new_num);
      ++new_num;
      operations.push_back('+');
    }
    s.pop();
    operations.push_back('-');
  }
  return operations;
}

void get_input(std::vector<int>* p_v) {
  int len;

  std::cin >> len;
  p_v->resize(len);
  for (int idx = 0; idx < len; ++idx)
    std::cin >> (*p_v)[idx];
}

int main(void) {
  std::vector<int>  nums;

  get_input(&nums);

  std::vector<char> operations = format_by_stack(nums);
  if (!operations.empty()) {
    for (size_t idx = 0; idx < operations.size(); ++idx)
      std::cout << operations[idx] << '\n';
  } else {
    std::cout << "NO";
  }
  return 0;
}

File: 1874/1874.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    1874.py                                            :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/08/04 15:10:05 by jihoolee          #+#    #+#              #
#    Updated: 2024/08/04 15:25:21 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

from collections import deque


def format_by_stack(nums: list[int]) -> list[str]:
    stack: deque[int] = deque()
    operations: list[str] = []
    new_num: int = 1

    for num in nums:
        while len(stack) == 0 or stack[-1] != num:
            if new_num > len(nums):
                return []
            stack.append(new_num)
            new_num += 1
            operations.append('+')
        stack.pop()
        operations.append('-')
    return operations


def main() -> None:
    nums: list[int] = [int(input()) for _ in range(int(input()))]
    operations: list[str] = format_by_stack(nums)
    if not operations:
        print('NO')
    else:
        print(*operations, sep='\n')


if __name__ == '__main__':
    main()
